File: python/integer_break.py
# no.343

class Solution:
    def product(self, nums):
        result = 1
        for i in nums:
            result *= i
        return result

    def integerBreak(self, n: int) -> int:
        best_sum = {1: [1], 2: [1, 1]}
        if n == 2:
            return 1
        def get_decomposed(nums):
            results = []
            for j in range(len(nums)):
                results.append(nums[:j] + best_sum[can[j]] + can[j+1:])
            return results

        for i in range(3, n+1):
            last_state = best_sum[i-1]
            max_product = 1
            for j in range(len(last_state)):
                can = last_state.copy()
                can[j] += 1
                if self.product(can) > max_product:
                    max_product = self.product(can)
                    selected = can
                    decomposed_can = get_decomposed(can)
                    for can in decomposed_can:
                        if self.product(can) >= max_product:
                            selected = can
                    candidate_list = selected
            best_sum[i] = candidate_list

        print(best_sum)

if __name__ == '__main__':
    sol = Solution()
    sol.integerBreak(58)

                    

File: C++/integer_break.cpp
// no.343
#include<iostream>
#include<vector>
#include<map>
using namespace std;

class Solution {
    public:
        map <int, vector<int> > best_sum;
        Solution(){
            best_sum[1].push_back(1);
            best_sum[2].push_back(1);
            best_sum[2].push_back(1);
        }
        vector<vector<int> > decompose(vector<int> v){
            vector<vector<int> >result;
            for (int i = 0; i < v.size(); i++){
                vector<int> r;
                int end = max(0, i - 2);
                r.insert(r.end(), v.begin(), v.begin() + end);
                r.insert(r.end(), best_sum[v[i]].begin(), best_sum[v[i]].end());
                r.insert(r.end(), v.begin() + i + 1, v.end());
                result.push_back(r);
                }
            return result;
            }

        int product(vector<int> v){
            int result = 1;
            for (int i = 0; i < v.size(); i++){
                result *= v[i];
            }
            return result;
        }

        int integerBreak(int n) {
            if (n == 2){
                return 1;
            }
            for (int i = 3; i <= n; i++){
                vector<int> last_state = best_sum[i - 1];
                int max_product = 1;
                vector<int> selected;
                for (int j = 0; j < last_state.size(); j++){
                    vector<int> candidate = last_state;
                    candidate[j] += 1;
                    int p = product(candidate);
                    if (p > max_product){
                        max_product = p;
                        selected = candidate;
                        vector<vector<int> > decomposed_candidate = decompose(candidate);
                        for (int k=0; k < decomposed_candidate.size(); k++){
                            int p = product(decomposed_candidate[k]);
                            if(p >= max_product){
                                selected = decomposed_candidate[k];
                                max_product = p;
                            }
                        }
                    }
                }
                best_sum[i] = selected;
            }
        return product(best_sum[n]);
    }
    void print_best_sum(){
        for(map<int, vector<int> >::iterator it = best_sum.begin(); it != best_sum.end(); it++){
            cout << it->first << ": ";
            print_vector(it->second);
            cout << endl;
        }

    }
    void print_vector(vector<int> v){
        cout << "[";
        for(int i = 0; i < v.size(); i++){
            cout << v[i] << " ";
        }
        cout << "]" << endl;
    }
};

int main(){
    Solution sol;
    cout << sol.integerBreak(58) << endl;
    sol.print_best_sum();
    return 1;
}
